File: client-web-api/php/lib/trisano-web-api.php
<?php

require_once('../lib/simpletest/browser.php');

class TrisanoWebApi {
  public $browser;
  private $base_url;
  public $page;

  function __construct() {
    if (!isset($_ENV['TRISANO_BASE_URL'])) die('Missing TRISANO_BASE_URL environment variable');
    $this->browser = &new SimpleBrowser();
    $this->browser->useCookies();
    $this->base_url = $_ENV['TRISANO_BASE_URL'];

    if (isset($_ENV['TRISANO_API_AUTH']) && strtolower($_ENV['TRISANO_API_AUTH']) != 'none') {
      $username = isset($_ENV['TRISANO_API_USER']) ? $_ENV['TRISANO_API_USER'] : die('Missing TRISANO_API_USER environment variable');
      $password = isset($_ENV['TRISANO_API_PASS']) ? $_ENV['TRISANO_API_PASS'] : die('Missing TRISANO_API_PASS environment variable');
    }

    if ($_ENV['TRISANO_API_AUTH'] == 'basic') {
     $this->browser->authenticate($username, $password);
    }
    elseif ($_ENV['TRISANO_API_AUTH'] == 'siteminder') {
      $auth_url = isset($_ENV['TRISANO_SM_URL']) ? $_ENV['TRISANO_SM_URL'] : die('Missing TRISANO_SM_URL environment variable');

      // Generate Siteminder cookies
      $this->browser->post($auth_url, array('LoginID' => $username,
                                            'Password' => $password,
                                            'Dispatch' => 'loginAuth'));
      if ($this->browser->getTransportError()) {
        die($this->browser->getTransportError());
      }
    }
  }

  public function get($path) {
    $this->page = $this->browser->get($this->base_url . $path);
    return $this->page;
  }

  public function get_page() {
    return $this->page;
  }

}

File: client-web-api/ruby/lib/trisano-web-api.rb
require 'rubygems'
require 'mechanize'

class TrisanoWebError < StandardError
  @errors = []

  attr_accessor :errors
  attr_accessor :response_code
end

class TriSanoWebApi 
  attr_accessor :agent

  def initialize
    @agent = Mechanize.new { |a|
      a.user_agent = 'TriSano Web API/1.0'
    }

    @base_url = ENV['TRISANO_BASE_URL'] || raise('Missing TRISANO_BASE_URL environment variable')
    url_comps = URI.parse(@base_url)
    @base_url_no_path = URI::Generic.new(url_comps.scheme, nil, url_comps.host, url_comps.port, nil, nil, nil, nil, nil).to_s

    if !ENV['TRISANO_API_AUTH'].nil? and ENV['TRISANO_API_AUTH'].downcase != 'none'
      username = ENV['TRISANO_API_USER'] || raise('Missing TRISANO_API_USER environment variable')
      password = ENV['TRISANO_API_PASS'] || raise('Missing TRISANO_API_PASS environment variable')
    end

    if ENV['TRISANO_API_AUTH'] == 'basic' then
      @agent.auth(username, password)
    elsif ENV['TRISANO_API_AUTH'] == 'siteminder' then
      auth_url = ENV['TRISANO_SM_URL'] || raise('Missing TRISANO_SM_URL environment variable')

      # Generate Siteminder cookies
      @agent.post(auth_url, { 'LoginID' => username,
                              'Password' => password,
                              'Dispatch' => 'loginAuth'});
    end
    
  end

  def home
    @agent.get(@base_url)
  end

  def get(uri, leading_path=true)
    @agent.get(base_uri(leading_path) + uri)
  end

  def post(uri, query, leading_path=true)
    http_action { @agent.post(base_uri(leading_path) + uri, query) }
  end

  def delete(uri, leading_path=true)
    http_action { @agent.delete(base_uri(leading_path) + uri) }
  end

  def put(uri, query, leading_path=true)
    http_action { @agent.put(base_uri(leading_path) + uri, query) }
  end

  def submit(form, button)
    http_action { @agent.submit(form, button) }
  end

  private

  def http_action
    begin
      yield
    rescue Mechanize::ResponseCodeError => response_error
      local_errors = []
      errors = response_error.page.search(".//div[@class = 'errorExplanation']")
      errors.each { |e|
        e.search(".//li").each { |detail|
          local_errors << detail.text.strip
        }
      }
      e = TrisanoWebError.new
      e.errors = local_errors
      e.response_code = response_error.response_code
      raise e
    end
  end

  def base_uri(leading_path)
    leading_path ? @base_url : @base_url_no_path
  end
end
